### PR TITLE
Add metadata to TypedArray and validate at runtime

### PR DESCRIPTION
**Pull request by raulsntos**
- Adds metadata to `ContainerTypeValidate` to validate at runtime that elements added to TypedArray fit the the element type.
- Adds metadata to TypedArray's `GetTypeInfo<>` so it's included in `extension_api.json` and language bindings can use the right element type for the Array.

As an example, the method `CodeEdit::get_folded_lines` that returns `TypedArray<int>`:

https://github.com/godotengine/godot/blob/c184105cf55614078b509b0a57c4bbd0beb62be7/scene/gui/code_edit.cpp#L1722

```diff
{
    "name": "get_folded_lines",
    "is_const": true,
    "is_vararg": false,
    "is_static": false,
    "is_virtual": false,
    "hash": 3995934104,
    "return_value": {
        "type": "typedarray::int",
+       "meta": "int32"
    }
},
```

### CODE REVIEW

**Comment by dsnopek** [comment]
Code-wise, I think this is the right approach to do these checks.

However, I think we should have a core discussion to decide if we _want_ to have these checks. Do we want `TypedArray` to be specific to integer and float sizes? Or, do we only want to deal with `Variant` types (and so it's always `int64_t` and `double`)?

I could personally see arguments for both sides.

Getting specific with sizes potentially allows packing the data tighter when copying to/from native containers (although, inside the `TypedArray`, it'll always be `Variant`s - to get packed data on both the Godot and native side the `Packed*Array` types are better). However, the added checks have an effect on performance on every set or get operation, for something that we might not really care about?

**Comment by raulsntos** [comment]
> I think we should have a core discussion to decide if we _want_ to have these checks.

That's fair. I said this on RC but I'll say it again here for the record, the motivation behind this PR is to generate GDExtension-based C# bindings with the correct element type for typed arrays.

The godot-cpp bindings generator is using the element type [as-is](https://github.com/godotengine/godot-cpp/blob/19c83a8837738e5014cc35771820bcb8cb73a5ea/binding_generator.py#L2779-L2780). So instead of assuming `int` means 64-bit like everywhere else where `int` appears, the generated APIs use `int` (not `int32_t` or `int64_t`).

I also checked what godot-rust does, and they do treat `int` like it means 64-bit so their generated APIs use typed arrays of `i64`.

And this is the list of APIs where a TypedArray with an integer element type appears (grouped by element type):

- APIs with `TypedArray<int>`:
    - [`CodeEdit::get_folded_lines`](https://github.com/godotengine/godot/blob/568589c9d8c763bfb3a4348174d53b42d7c59f21/scene/gui/code_edit.cpp#L1722)
    - [`CodeEdit::set_line_length_guidelines`](https://github.com/godotengine/godot/blob/568589c9d8c763bfb3a4348174d53b42d7c59f21/scene/gui/code_edit.cpp#L2285)
    - [`CodeEdit::get_line_length_guidelines`](https://github.com/godotengine/godot/blob/568589c9d8c763bfb3a4348174d53b42d7c59f21/scene/gui/code_edit.cpp#L2290)
    - [`Input::get_connected_joypads`](https://github.com/godotengine/godot/blob/568589c9d8c763bfb3a4348174d53b42d7c59f21/core/input/input.cpp#L1644)
- APIs with `TypedArray<uint64_t>`:
    - [`PhysicsTestMotionParameters2D::get_exclude_objects`](https://github.com/godotengine/godot/blob/568589c9d8c763bfb3a4348174d53b42d7c59f21/servers/physics_server_2d.cpp#L485)
    - [`PhysicsTestMotionParameters2D::set_exclude_objects`](https://github.com/godotengine/godot/blob/568589c9d8c763bfb3a4348174d53b42d7c59f21/servers/physics_server_2d.cpp#L497)
    - [`PhysicsTestMotionParameters3D::get_exclude_objects`](https://github.com/godotengine/godot/blob/568589c9d8c763bfb3a4348174d53b42d7c59f21/servers/physics_server_3d.cpp#L512)
    - [`PhysicsTestMotionParameters3D::set_exclude_objects`](https://github.com/godotengine/godot/blob/568589c9d8c763bfb3a4348174d53b42d7c59f21/servers/physics_server_3d.cpp#L524)
- APIs with `TypedArray<int64_t>`:
    - [`RDPipelineMultisampleState::set_sample_masks`](https://github.com/godotengine/godot/blob/568589c9d8c763bfb3a4348174d53b42d7c59f21/servers/rendering/rendering_device_binds.h#L567)
    - [`RDPipelineMultisampleState::get_sample_masks`](https://github.com/godotengine/godot/blob/568589c9d8c763bfb3a4348174d53b42d7c59f21/servers/rendering/rendering_device_binds.h#L568)

**Comment by Bromeon** [comment]
> However, the added checks have an effect on performance on every set or get operation, for something that we might not really care about?

If static type information about the array is available (`Array[T]`), it should only be necessary to add the checks for the following element types (when converting from `Variant`):
- `int8_t`
- `int16_t`
- `int32_t`
- `uint8_t`
- `uint16_t`
- `uint32_t`

If the array is untyped (`Array`), you'd still need to query the `Variant` type, but this is already done today. You would only need to do an additional check (type of inserted value vs. element meta) for the above types.

There are some special cases to consider:
- `int64_t` -- this is the native integer type. Since all its values are perfectly representable, `Array<int64_t>` wouldn't need a runtime check.
- `uint64_t` -- it is arguable whether this type should be supported. Not all its values can be stored as `int64_t`, so if you `TypedArray<uint64_t>` to an untyped array, you'll end up with _different_ values as elements, which can introduce subtle bugs.
- `float` and `double`. Since the native `Variant` type is `double`, it should be able to hold all values of both types. If the user decides to use `float` for inserting/extracting elements, should that really be an error?


---

**TLDR:** unless I'm missing something, using `TypedArray<T>` with `T = int64_t, float, double` should not incur performance penalty.
